lib/src/main/java/org/kpipe/consumer/OffsetManager.java

```diff
@@ -0,0 +1,574 @@
+package org.kpipe.consumer;
+
+import static java.util.logging.Level.WARNING;
+
+import java.time.Duration;
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.logging.Logger;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.kpipe.consumer.enums.OffsetState;
+
+/**
+ * Manages Kafka consumer offsets for parallel processing scenarios.
+ *
+ * <p>This class provides safe offset commit management by tracking which offsets are being
+ * processed and which have completed processing. It ensures that only offsets that have been fully
+ * processed are committed, even when processing happens in a non-sequential order.
+ *
+ * <p>Key features include:
+ *
+ * <ul>
+ *   <li>Safe tracking of offsets being processed in parallel
+ *   <li>Support for non-contiguous offset processing
+ *   <li>Thread-safe offset management
+ *   <li>Proper offset commit logic to avoid data loss
+ *   <li>Automatic cleanup of processed offsets to prevent memory leaks
+ *   <li>Periodic offset commits with configurable interval
+ *   <li>Efficient execution using virtual threads for I/O operations
+ *   <li>Diagnostic utilities for monitoring and troubleshooting
+ * </ul>
+ *
+ * @param <K> The type of the key
+ * @param <V> The type of the value
+ */
+public class OffsetManager<K, V> implements AutoCloseable {
+
+  private static final Logger LOGGER = Logger.getLogger(OffsetManager.class.getName());
+  private static final long DEFAULT_COMMAND_TIMEOUT_MS = 5000;
+
+  private final Consumer<K, V> consumer;
+  private final ExecutorService executorService;
+  private final AtomicReference<OffsetState> state = new AtomicReference<>(OffsetState.CREATED);
+
+  // Tracking data structures
+  private final Map<TopicPartition, ConcurrentSkipListSet<Long>> pendingOffsets = new ConcurrentHashMap<>();
+  private final Map<TopicPartition, ConcurrentSkipListSet<Long>> processedOffsets = new ConcurrentHashMap<>();
```

Comment:
I am not really convinced that both maps are needed. 

I think it would be a bit simpler to just maintain pending (inflight) offsets and nextOffsetsToCommit map.

On processing completion for the offset:
- remove from pending offsets
- set nextOffsetsToCommit if processed offset is above current value in nextOffsetsToCommit (would rename to highestProcessedOffset)

On commit - just commit up to Min(lowest pending, highestProcessedOffset +1 ) - so if there are no pending - just commit to highestProcessed+1, but if there are pending / inflight - commit to first / lowest inflight.

Reply:
agreed and implmented